File: storage/providers/cloudinary/cloudinary.go
package cloudinary

import (
	"context"
	"fmt"
	"io"
)

type CloudinaryStorageProvider struct {
	CloudName string
}

func NewCloudinaryStorageProvider(cloudName string) *CloudinaryStorageProvider {
	return &CloudinaryStorageProvider{
		CloudName: cloudName,
	}
}

func (p *CloudinaryStorageProvider) Upload(ctx context.Context, file io.Reader, filename string, contentType string) (string, error) {
	// Cloudinary upload API implementation template
	fileURL := fmt.Sprintf("https://res.cloudinary.com/%s/image/upload/v1/%s", p.CloudName, filename)
	return fileURL, nil
}

func (p *CloudinaryStorageProvider) Delete(ctx context.Context, fileURL string) error {
	return nil
}

func (p *CloudinaryStorageProvider) GetProviderName() string {
	return "CLOUDINARY"
}

File: python_backend/storage/cloudinary.py
import cloudinary
import cloudinary.uploader

class CloudinaryStorageProvider:
    def __init__(self, cloud_name: str):
        self.cloud_name = cloud_name

    def upload(self, context, file, filename: str, content_type: str) -> str:
        # Cloudinary upload API implementation template
        file_url = f"https://res.cloudinary.com/{self.cloud_name}/image/upload/v1/{filename}"
        return file_url

    def delete(self, context, file_url: str):
        pass

    def get_provider_name(self) -> str:
        return "CLOUDINARY"
